Add per-facilitator questions to the PD Workshop Survey

Hamms opened the pull request:
Adds a React component to the PD Workshop Survey form that dynamically creates a table giving certain questions on a facilitator-by-facilitator basis:

![facilitator](https://cloud.githubusercontent.com/assets/244100/24064376/529d22f6-0b21-11e7-9f12-731c0c6f3a5a.gif)

And also a series of free response questions:

![facilitator](https://cloud.githubusercontent.com/assets/244100/24064504/13327b9c-0b22-11e7-9c3f-17b0319d0667.gif)

Also validates the responses server-side to require at least one facilitator, and a radio selected for every question for every facilitator selected.

Also updates the summary report to handle the per-facilitator questions appropriately; facilitator reports will reflect their individual responses, and overview reports will reflect an averaging of facilitator-by-facilitator responses.

This PR is unfortunately rather extensive, but each commit should represent a focused subset of work; I would recommend reviewing commit-by-commit.

Hamms commented on apps/src/code-studio/pd/workshop_survey/VariableFormGroup.jsx:
```diff
@@ -0,0 +1,148 @@
+      formGroup = <RadioFormGroup name={key} values={this.props.question.values} />;
+    } else if (this.props.question.type === MULTI_SELECT) {
+      formGroup = <CheckboxFormGroup name={key} values={this.props.question.values} />;
+    }
```
Note that right now, columns only support single- or multi-select options, and rows only support free response. This is slightly more than we need for the current implementation, but both could reasonably be expanded to support more types for future needs.

mehalshah commented on apps/src/code-studio/pd/workshop_survey/CheckboxFormGroup.jsx:
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
```
I'm a little surprised that you have to reimplement this. There's a control used in teacher_application.jsx that you might use.

Hamms replied:
replaced!

mehalshah commented on dashboard/lib/api/v1/pd/workshop_score_summarizer.rb:
```diff
@@ -96,30 +107,15 @@ def get_score_for_workshops(workshops, facilitator_breakdown: false, include_fre
-            # Do nothing - no score to compute but don't skip this
+        survey_response.symbolize_keys.each do |question, answer|
+          if answer.is_a? Hash
+            # Then "answer" is actually a hash af answers for each
```
af -> of

Hamms replied:
done

mehalshah commented on pegasus/forms/pd_workshop_survey.rb:
```diff
@@ -33,15 +33,30 @@ def self.normalize(data)
-      result[:things_facilitator_could_improve_s] = stripped data[:things_facilitator_could_improve_s]
+      result[:who_facilitated_ss] = required data[:who_facilitated_ss]
+
+      unless result[:who_facilitated_ss].class == FieldError
```
Why is this a FieldError? Shouldn't it be blank?

Hamms replied:
No, because of line 37; since this is a required field, leaving it blank counts as a FieldError

mehalshah commented on pegasus/test/test_forms_pd_workshop_survey.rb:
```diff
@@ -0,0 +1,133 @@
+        {
+          consent_b: '1',
+          who_facilitated_ss: facilitators,
+          how_clearly_presented_s: {"Tom" => 'Extremely clearly', "Dick" => 'Extremely clearly'},
```
Vary these answers

Hamms replied:
done

Hamms commented:
Updated clientside code to use ButtonList, and removed the new custom form elements.

Also moved the asterisk:

![image](https://cloud.githubusercontent.com/assets/244100/24266341/a9e6da60-0fc3-11e7-8b9c-ac68857d838b.png)

PTAL

mehalshah commented on apps/src/code-studio/pd/form_components/button_list.jsx:
```diff
@@ -1,6 +1,12 @@
 import React from 'react';
 import _ from 'lodash';
-import {Radio, Checkbox, ControlLabel, FormGroup, HelpBlock} from 'react-bootstrap';
+import {
```
Is this a style preference anywhere? If so, should we be enforcing this somewhere?

Hamms replied:
[it is](https://github.com/airbnb/javascript#modules--multiline-imports-over-newlines), but I unfortunately don't think eslint has an option to complain about this

mehalshah commented on apps/src/code-studio/pd/form_components/button_list.jsx:
```diff
@@ -20,7 +26,6 @@ const ButtonList = React.createClass({
 
   handleChange(event) {
     let value;
-    console.log(event);
```
Maybe we should have a similar rule for console.log

Hamms replied:
We absolutely should! looks like we currently have over a hundred violations of [no-console](http://eslint.org/docs/rules/no-console), so it at the very least warrants its own PR.

mehalshah commented on apps/src/code-studio/pd/workshop_survey/VariableFormGroup.jsx:
```diff
@@ -0,0 +1,158 @@
+      />
+    );
+
+    return (<td key={key}>{formGroup}</td>);
```
Should this be multiline as well?

Hamms replied:
[nope](https://github.com/airbnb/javascript/tree/master/react#alignment); short components are allowed to stay on a single line

Hamms commented:
Tests are already written in a separate PR: https://github.com/code-dot-org/code-dot-org/pull/13894
